google-cloud-firestore/src/main/java/com/google/cloud/firestore/telemetry/MetricsUtil.java

```diff
@@ -0,0 +1,254 @@
+import com.google.api.core.ApiFutures;
+import com.google.api.gax.rpc.StatusCode;
+import com.google.api.gax.tracing.ApiTracerFactory;
+import com.google.api.gax.tracing.MetricsTracerFactory;
+import com.google.api.gax.tracing.OpenTelemetryMetricsRecorder;
+import com.google.cloud.firestore.FirestoreException;
+import com.google.cloud.firestore.FirestoreOptions;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Stopwatch;
+import com.google.common.util.concurrent.MoreExecutors;
+import io.grpc.Status;
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.api.metrics.DoubleHistogram;
+import io.opentelemetry.api.metrics.Meter;
+import io.opentelemetry.sdk.OpenTelemetrySdk;
+import io.opentelemetry.sdk.metrics.SdkMeterProvider;
+import io.opentelemetry.sdk.metrics.SdkMeterProviderBuilder;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * A utility interface for trace collection. Classes that implement this interface may make their
+ * own design choices for how they approach trace collection. For instance, they may be no-op, or
+ * they may use a particular tracing framework such as OpenTelemetry.
+ */
+public class MetricsUtil {
+
+  private final FirestoreOptions firestoreOptions;
+  private OpenTelemetry openTelemetry;
+  private ApiTracerFactory otelApiTracerFactory;
+
+  private Meter meter;
+  private DoubleHistogram endToEndRequestLatency;
+  private DoubleHistogram firstResponseLatency;
+
+  /**
+   * Creates and returns an instance of the MetricsUtil class.
+   *
+   * @param firestoreOptions The FirestoreOptions object that is requesting an instance of
+   *     MetricsUtil.
+   * @return An instance of the MetricsUtil class.
+   */
+  public MetricsUtil(FirestoreOptions firestoreOptions) {
+    this.firestoreOptions = firestoreOptions;
+
+    // TODO(mila): this checking should be done inside FirestoreOptions builder, so that we can
+    // setApiTracerFactory
+    // to collect GAX layer metrics
+    boolean createEnabledInstance = firestoreOptions.getOpenTelemetryOptions().isMetricsEnabled();
+
+    // The environment variable can override options to enable/disable telemetry collection.
+    String enableMetricsEnvVar = System.getenv(ENABLE_METRICS_ENV_VAR);
+    if (enableMetricsEnvVar != null) {
+      if (enableMetricsEnvVar.equalsIgnoreCase("true")
+          || enableMetricsEnvVar.equalsIgnoreCase("on")) {
+        createEnabledInstance = true;
+      }
+      if (enableMetricsEnvVar.equalsIgnoreCase("false")
+          || enableMetricsEnvVar.equalsIgnoreCase("off")) {
+        createEnabledInstance = false;
+      }
+    }
+
+    // TODO(mila): re-assess if it should follow tracing's design: enabled/disabled MetricsUtil
```

Comment:
I know I mentioned metrics doesn't necessarily need to follow the enabled/disable MetricsUtil design. Now that I have read the code I have a better sense of the question, and I think we should probably do the enabled/disabled classes. Let's discuss it tomorrow.